docs/manual/scala/guide/cluster/DomainModelling.md

```diff
@@ -1,19 +1,202 @@
+
+A simplified shopping cart example is used to guide you through the process. You can find a full-fledge shopping cart sample on our [samples repository](https://github.com/lagom/lagom-samples/tree/1.6.x/shopping-cart/shopping-cart-scala).
 
 ## Encoding the model
-  * typed protocol - Command, Event and State
-  * defining typed replies and error handling
-  * Enforced replies
-  * Changing behavior - FSM
-## EventSourcingBehaviour - glueing the bits together
-  * Tagging the events - Akka Persistence Query considerations
-  * Configuring snaptshots
+
+Start by defining your model in terms of Commands, Events, and State.
+
+### Modelling the State
+
+The state of the shopping cart is defined as following:
+
+```scala
+/** Common Shopping Cart trait */
+sealed trait ShoppingCart
+/** Once we add new items it becomes an open shopping cart */
+case class OpenShoppingCart(items: Map[String, Int]) extends ShoppingCart
+/** Once checked-out it reaches end-of-life and can't be used anymore */
+case class CheckedOutShoppingCart(items: Map[String, Int]) extends ShoppingCart
+```
+
+Note that we are modelling it using a common trait `ShoppingCart` and two possible states modelled as extensions of `ShoppingCart`. This is a recommended technique whenever your model go through different state transitions. As we will see later, each state encodes which commands it can handle, which events it can persist and to which other states it can transition.
+
+### Modelling Commands and Replies
+
+Next we define the commands that we can send to it.
+
+Each command defines a [reply](https://doc.akka.io/docs/akka/2.6/typed/persistence.html#replies) through a `replyTo: ActorRef[R]` field where `R` is the reply that will be sent back to the caller. Replies are used to communicate back if a command was accepted or rejected or to read the aggregate data (ie: read-only commands). It's also possible to have a mix of both, for example, if the command succeeds it returns some updated data; if it fails it returns a rejected message. Or you can have commands without replies (ie: fire-and-forget). This is a less common pattern in CQRS Aggregate modeling though and not covered in this example.
+
+In Akka Typed, it's not possible to return an exception to the caller. All communication between the actor and the caller must be done via the `replyTo:ActorRef[R]` passed in the command. Therefore, if you want to signal a rejection, you most have it encoded in your reply protocol.
+
+```scala
+// Replies
+sealed trait Confirmation
+sealed trait Accepted extends Confirmation
+case object Accepted extends Accepted
```

Comment:
(future improvement)

I remember you taught me this idiom and the reason (avoid `Accepted.type`) but I'm not sure all readers will be familiar with it and it could raise some questions. It'd be great to add a link to some blogpost/docs about why `trait` + `object` is useful.

Reply:
I can re-think this a little further. I remember to have hit an issue with play-json serializers once. I don't remember what it was. It's possible that we make it a class for the sake of simplicity. If we need to add a link or explain why, maybe we can just avoid the issue.